src/main/io/rcdevice.c

```diff
@@ -0,0 +1,1013 @@
+                            isFoundANullTeminatedChar = true; 
+                            expecteRemainingDataLen = 1;
+                        }
+                    }
+                } else if (settingType == RCDEVICE_PROTOCOL_SETTINGTYPE_STRING) {
+                    // found the null teminated char, and get max string size and crc field, then break the loop
+                    isFoundANullTeminatedChar = true; 
+                    expecteRemainingDataLen = 2;
+                }
+            }
+
+            if (dataPos >= expectedPacketLen) { 
+                break;
+            }
+        }
+    }
+
+    if (crc != 0) {
+        return false;
+    }
+
+    if (outputBufferLen && (*outputBufferLen >= (dataPos - 1)) && outputBuffer) {
+        memcpy(outputBuffer, data,
+               dataPos - 1); // return the data expect the crc field
+        *outputBufferLen = dataPos - 1;
+    }
+
+    return true;
+}
+
+// every time send packet to device, and want to get something from device,
+// it'd better call the method to clear the rx buffer before the packet send,
+// else may be the useless data in rx buffer will cause the response decoding
+// failed.
+static void runcamDeviceFlushRxBuffer(runcamDevice_t *device)
+{
+    while (serialRxBytesWaiting(device->serialPort) > 0) {
+        serialRead(device->serialPort);
+    }
+}
+
+// a common way to send packet to device
+static void runcamDeviceSendPacket(runcamDevice_t *device, uint8_t command, uint8_t *paramData, int paramDataLen)
+{
+    // is this device open?
+    if (!device->serialPort) {
+        return;
+    }
+
+    sbuf_t buf;
+    // prepare pointer
+    buf.ptr = device->buffer;
+    buf.end = ARRAYEND(device->buffer);
+
+    sbufWriteU8(&buf, RCDEVICE_PROTOCOL_HEADER);
+    sbufWriteU8(&buf, command);
+
+    if (paramData) {
+        sbufWriteData(&buf, paramData, paramDataLen);
+    }
+
+    // add crc over (all) data
+    crc8_dvb_s2_sbuf_append(&buf, device->buffer);
+
+    // switch to reader
+    sbufSwitchToReader(&buf, device->buffer);
+
+    // send data if possible
+    serialPort_t *serialPort = device->serialPort;
+    if (!serialPort->locked) {
+        serialPort->locked = true;
+        serialWriteBuf(device->serialPort, sbufPtr(&buf), sbufBytesRemaining(&buf));
+        serialPort->locked = false;
+    }
+}
+
+// a common way to receive data from device, and will decode the response,
+// and save the response data(all the fields from response except the header and
+// crc field) to outputBuffer if the outputBuffer not a NULL pointer.
+static bool runcamDeviceSerialReceive(runcamDevice_t *device, uint8_t command, uint8_t *outputBuffer, uint8_t *outputBufferLen)
+{
+    // wait 1000ms for reply
+    bool headerReceived = false;
+    timeMs_t timeout = millis() + 1000;
+    while (millis() < timeout) {
+        if (serialRxBytesWaiting(device->serialPort) > 0) {
+            if (!headerReceived) {
```

Comment:
Double negative. Change to `if (headerReceived) {` and rearrange `if` and `else` blocks accordingly.